src/main/scala/org/broadinstitute/dsde/agora/server/business/PermissionBusiness.scala

```diff
@@ -1,72 +1,114 @@
     AgoraEntityPermissionsClient.listEntityPermissions(entity)
   }
 
-  def insertEntityPermission(entity: AgoraEntity, requester: String, accessObject: AccessControl) = {
-    authorizeEntityRequester(entity, requester)
+  def insertEntityPermission(entity: AgoraEntity, requester: String, accessObject: AccessControl): Int = {
+    // Inserts can result in an edit in the DAO layer
+    val entityACLs = getEntityACLs(entity, requester)
+    checkSameRequesterAndPermissions(entityACLs.find(acl => acl.user.equals(requester)), accessObject)
+    authorizeEntityRequester(entityACLs, entity, requester)
     AgoraEntityPermissionsClient.insertEntityPermission(entity, accessObject)
   }
 
-  def batchEntityPermission(entity: AgoraEntity, requester: String, accessObjectList: List[AccessControl]) = {
+  def batchEntityPermission(entity: AgoraEntity, requester: String, accessObjectList: List[AccessControl]): Int = {
     accessObjectList.map(accessObject => insertEntityPermission(entity, requester, accessObject)).sum
   }
 
-  def editEntityPermission(entity: AgoraEntity, requester: String, accessObject: AccessControl) = {
-    authorizeEntityRequester(entity, requester)
+  def editEntityPermission(entity: AgoraEntity, requester: String, accessObject: AccessControl): Int = {
+    val entityACLs = getEntityACLs(entity, requester)
+    checkSameRequesterAndPermissions(entityACLs.find(acl => acl.user.equals(requester)), accessObject)
+    authorizeEntityRequester(entityACLs, entity, requester)
     AgoraEntityPermissionsClient.editEntityPermission(entity, accessObject)
 
   }
 
-  def deleteEntityPermission(entity: AgoraEntity, requester: String, userToRemove: String) = {
-    authorizeEntityRequester(entity, requester)
+  def deleteEntityPermission(entity: AgoraEntity, requester: String, userToRemove: String): Int = {
+    checkSameRequester(requester, userToRemove)
+    authorizeEntityRequester(getEntityACLs(entity, requester), entity, requester)
     AgoraEntityPermissionsClient.deleteEntityPermission(entity, userToRemove)
   }
 
-  def authorizeNamespaceRequester(entity: AgoraEntity, requester: String) = {
-    if (!NamespacePermissionsClient.getNamespacePermission(entity, requester).canManage &&
-        !NamespacePermissionsClient.getNamespacePermission(entity, "public").canManage)
-      throw new NamespaceAuthorizationException(AgoraPermissions(Manage), entity, requester)
+
+  private def authorizeNamespaceRequester(acls: Seq[AccessControl], entity: AgoraEntity, requester: String): Unit = {
+    if (!acls.exists(_.roles.canManage))
+      throw NamespaceAuthorizationException(AgoraPermissions(Manage), entity, requester)
+  }
+
+  private def authorizeEntityRequester(acls: Seq[AccessControl], entity: AgoraEntity, requester: String): Unit = {
+    if (!acls.exists(_.roles.canManage))
+      throw AgoraEntityAuthorizationException(AgoraPermissions(Manage), entity, requester)
+  }
```

Comment:
could you remove some of the boilerplate from the business methods above (e.g. `insertEntityPermission`) and move it down into these private utility methods? For instance, if you had:
```
  private def authorizeEntityRequester(entity: AgoraEntity, requester: String): Unit = {
    authorizeEntityRequester(getEntityACLs(entity, requester), entity, requester)
  }

  private def authorizeEntityRequester(entity: AgoraEntity, requester: String, accessObject: AccessControl): Unit = {
    val currentACLs = getEntityACLs(entity, requester)
    checkSameRequesterAndPermissions(currentACLs.find(acl => acl.user.equals(requester)), accessObject)
    authorizeEntityRequester(currentACLs, entity, requester)
  }
```

then a bunch of the code up above could be simpler.